File: tempo/resample.py
import pyspark.sql.functions as f
import tempo.tsdf as ts

# define global frequency options
SEC = 'sec'
MIN = 'min'
HR = 'hr'

# define global aggregate function options for downsampling
CLOSEST_LEAD = "closest_lead"
MIN_LEAD = "min_lead"
MAX_LEAD = "max_lead"
MEAN_LEAD = "mean_lead"

allowableFreqs = [SEC, MIN, HR]

def __appendAggKey(tsdf, freq = None):
    """
    :param tsdf: TSDF object as input
    :param freq: frequency at which to upsample
    :return: return a TSDF with a new aggregate key (called agg_key)
    """
    df = tsdf.df
    checkAllowableFreq(freq)

    # compute timestamp columns
    sec_col = f.second(f.col(tsdf.ts_col))
    min_col = f.minute(f.col(tsdf.ts_col))
    hour_col = f.hour(f.col(tsdf.ts_col))

    if (freq == SEC):
        #agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lpad(hour_col, 2, '0'), f.lpad(min_col, 2, '0'), f.lpad(sec_col, 2, '0'))
        agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lit(" "), f.lpad(hour_col, 2, '0'), f.lit(':'), f.lpad(min_col, 2, '0'), f.lit(':'), f.lpad(sec_col, 2, '0')).cast("timestamp")
    elif (freq == MIN):
        #agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lpad(hour_col, 2, '0'), f.lpad(min_col, 2, '0'))
        agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lit(' '), f.lpad(hour_col, 2, '0'), f.lit(':'), f.lpad(min_col, 2, '0'), f.lit(':'), f.lit('00')).cast("timestamp")
    elif (freq == HR):
        #agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lpad(hour_col, 2, '0'))
        agg_key = f.concat(f.col(tsdf.ts_col).cast("date"), f.lit(' '), f.lpad(hour_col, 2, '0'), f.lit(':'), f.lit('00'), f.lit(':'), f.lit('00')).cast("timestamp")

    df = df.withColumn("agg_key", agg_key)
    return ts.TSDF(df, tsdf.ts_col, partition_cols = tsdf.partitionCols)

def aggregate(tsdf, freq, func, metricCols = None):
    """
    aggregate a data frame by a coarser timestamp than the initial TSDF ts_col
    :param tsdf: input TSDF object
    :param func: aggregate function
    :param metricCols: columns used for aggregates
    :return: TSDF object with newly aggregated timestamp as ts_col with aggregated values
    """
    tsdf = __appendAggKey(tsdf, freq)
    df = tsdf.df

    groupingCols = tsdf.partitionCols + ['agg_key']
    if metricCols is None:
        metricCols = list(set(df.columns).difference(set(groupingCols + [tsdf.ts_col])))

    groupingCols = [f.col(column) for column in groupingCols]

    if func == CLOSEST_LEAD:
        #exprs = {x: "min" for x in metricCols}
        metricCol = f.struct([tsdf.ts_col] + metricCols)
        res = df.withColumn("struct_cols", metricCol).groupBy(groupingCols)
        res = res.agg(f.min('struct_cols').alias("closest_data")).select(*groupingCols, f.col("closest_data.*"))
    elif func == MEAN_LEAD:
        exprs = {x: "avg" for x in metricCols}
        res= df.groupBy(groupingCols).agg(exprs)
    elif func == MIN_LEAD:
        exprs = {x: "min" for x in metricCols}
        summaryCols = metricCols
        res = df.groupBy(groupingCols).agg(exprs)
    elif func == MAX_LEAD:
        exprs = {x: "max" for x in metricCols}
        summaryCols = metricCols
        res = df.groupBy(groupingCols).agg(exprs)

    res = res.drop(tsdf.ts_col).withColumnRenamed('agg_key', tsdf.ts_col)
    return(ts.TSDF(res, ts_col = tsdf.ts_col, partition_cols = tsdf.partitionCols))


def checkAllowableFreq(freq):
    if freq not in allowableFreqs:
      raise ValueError("Allowable grouping frequencies are sec (second), min (minute), hr (hour)")

def validateFuncExists(func):
  if func is None:
      raise ValueError("Aggregate function missing. Provide one of the allowable functions: " + ", ".join([CLOSEST_LEAD, MIN_LEAD, MAX_LEAD, MEAN_LEAD]))

File: scala/tempo/src/main/scala/com/databrickslabs/tempo/resample.scala
package com.databrickslabs.tempo

import org.apache.spark.sql.functions._
import TSDF._

/**
  * The following object contains methods for resampling (up to millions of time series in parallel).
  */
object resample {

// define global frequency options
val SEC = "sec"
val MIN = "min"
val HR = "hr"

// define global aggregate function options for downsampling
val CLOSEST_LEAD = "closest_lead"
val MIN_LEAD = "min_lead"
val MAX_LEAD = "max_lead"
val MEAN_LEAD = "mean_lead"

val allowableFreqs = List(SEC, MIN, HR)

  /**
    *
    * @param tsdf - TSDF object as input
    * @param freq - frequency at which to upsample
    * @return - return a TSDF with a new aggregate key (called agg_key)
    */
  private[tempo] def __appendAggKey(tsdf : TSDF, freq : String) : TSDF = {
    var df = tsdf.df
    checkAllowableFreq(freq)

    // compute timestamp columns
    val sec_col = second(col(tsdf.tsColumn.name))
    val min_col = minute(col(tsdf.tsColumn.name))
    val hour_col = hour(col(tsdf.tsColumn.name))
    var agg_key = col(tsdf.tsColumn.name)

    if (freq == SEC) {
      agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lpad(min_col, 2, "0"), lit(":"), lpad(sec_col, 2, "0")).cast("timestamp")
    } else if (freq == MIN) {
        agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lpad(min_col, 2, "0"), lit(":"), lit("00")).cast("timestamp")
    } else if (freq == HR) {
        agg_key = concat(col(tsdf.tsColumn.name).cast("date"), lit(" "), lpad(hour_col, 2, "0"), lit(":"), lit("00"), lit(":"), lit("00")).cast("timestamp")
    }

    df = df.withColumn("agg_key", agg_key)
    return TSDF(df, tsColumnName= tsdf.tsColumn.name, partitionColumnNames = tsdf.partitionCols.map(x => x.name): _*)
  }

  /**
    *
    * @param tsdf - input TSDF object
    * @param freq - aggregate function
    * @param func - aggregate function
    * @param metricCols - columns used for aggregates
    * return - TSDF object with newly aggregated timestamp as ts_col with aggregated values
    */
  def rs_agg(tsdf : TSDF, freq : String, func : String, metricCols : Option[List[String]] = None) : TSDF =  {

       validateFuncExists(func)

       var tsdf_w_aggkey = __appendAggKey(tsdf, freq)
       val df = tsdf_w_aggkey.df
       var adjustedMetricCols = List("")

       var groupingCols = tsdf.partitionCols.map(x => x.name) :+ "agg_key"

       adjustedMetricCols = metricCols match {
         case Some(s) => metricCols.get
         case None => (df.columns.toSeq diff (groupingCols :+ tsdf.tsColumn.name)).toList
       }


       var metricCol = col("")
       var groupedRes = df.withColumn("struct_cols", struct( (Seq(tsdf.tsColumn.name) ++ adjustedMetricCols.toSeq).map(col): _*)).groupBy(groupingCols map col: _*)
       var res = groupedRes.agg(min("struct_cols").alias("closest_data")).select("*", "closest_data.*").drop("closest_data")

       if (func == CLOSEST_LEAD)  {
         groupedRes = df.withColumn("struct_cols", struct( (Seq(tsdf.tsColumn.name) ++ adjustedMetricCols.toSeq).map(col): _*)).groupBy(groupingCols map col: _*)
         res = groupedRes.agg(min("struct_cols").alias("closest_data")).select("*", "closest_data.*").drop("closest_data")
       } else if (func == MEAN_LEAD) {
         val exprs = adjustedMetricCols.map(k => avg(df(s"$k")).alias("avg_" + s"$k"))
         res = df.groupBy(groupingCols map col: _*).agg(exprs.head, exprs.tail:_*)
       } else if (func == MIN_LEAD) {
         val exprs = adjustedMetricCols.map(k => min(df(s"$k")).alias("avg_" + s"$k"))
         res = df.groupBy(groupingCols map col: _*).agg(exprs.head, exprs.tail:_*)
       } else if (func == MAX_LEAD) {
         val exprs = adjustedMetricCols.map(k => max(df(s"$k")).alias("avg_" + s"$k"))
         res = df.groupBy(groupingCols map col: _*).agg(exprs.head, exprs.tail:_*)
       }

       res = res.drop(tsdf.tsColumn.name).withColumnRenamed("agg_key", tsdf.tsColumn.name)
       return(TSDF(res, tsColumnName = tsdf.tsColumn.name, partitionColumnNames = tsdf.partitionCols.map(x => x.name): _*))
  }


       def checkAllowableFreq(freq : String) : Unit = {
         if (!allowableFreqs.contains(freq) ) {
         throw new IllegalArgumentException ("Allowable grouping frequencies are sec (second), min (minute), hr (hour)")
       } else {}
       }


       def validateFuncExists(func : String): Unit = {
         if (func == None) {
           throw new IllegalArgumentException("Aggregate function missing. Provide one of the allowable functions: " + List(CLOSEST_LEAD, MIN_LEAD, MAX_LEAD, MEAN_LEAD).mkString(","))

         }
       }
  }